server/neptune/workflows/internal/deploy/revision/queue/revision_processor.go

```diff
@@ -0,0 +1,114 @@
+)
+
+func (p *RevisionProcessor) Process(ctx workflow.Context, requestedDeployment terraform.DeploymentInfo, latestDeployment *root.DeploymentInfo) error {
+	commitDirection, err := p.getDeployRequestCommitDirection(ctx, requestedDeployment, latestDeployment)
+	if err != nil {
+		return err
+	}
+
+	// TODO: remove log? it might be noisy
+	logger.Info(ctx, fmt.Sprintf("relationship of deployed to requested revision: %s", commitDirection),
+		"deployed-revision", latestDeployment.GetRevision(),
+		"requested-revision", requestedDeployment.Revision)
+
+	switch commitDirection {
+	case activities.DirectionBehind:
+		p.updateCheckRun(ctx, requestedDeployment, github.CheckRunFailure, DirectionBehindSummary)
+		return fmt.Errorf("requested revision %s is behind current one %s", requestedDeployment.Revision, latestDeployment.GetRevision())
+	case activities.DirectionDiverged:
+		return p.lock(ctx, requestedDeployment)
+	}
+	return nil
+}
+
+func (p *RevisionProcessor) getDeployRequestCommitDirection(ctx workflow.Context, deployRequest terraform.DeploymentInfo, latestDeployment *root.DeploymentInfo) (activities.DiffDirection, error) {
+	// root being deployed for the first time
+	if latestDeployment == nil {
+		return activities.DirectionAhead, nil
+	}
+
+	var compareCommitResp activities.CompareCommitResponse
+	err := workflow.ExecuteActivity(ctx, p.Activities.CompareCommit, activities.CompareCommitRequest{
+		DeployRequestRevision:  deployRequest.Revision,
+		LatestDeployedRevision: latestDeployment.Revision,
+		Repo:                   deployRequest.Repo,
+	}).Get(ctx, &compareCommitResp)
+	if err != nil {
+		return "", errors.Wrap(err, "unable to determine deploy request commit direction")
+	}
+	return compareCommitResp.CommitComparison, nil
+}
+
+// worker should not block on updating check runs for invalid deploy requests so let's retry for UpdateCheckrunRetryCount only
+func (p *RevisionProcessor) updateCheckRun(ctx workflow.Context, deployRequest terraform.DeploymentInfo, state github.CheckRunState, summary string) {
+	ctx = workflow.WithRetryPolicy(ctx, temporal.RetryPolicy{
+		MaximumAttempts: UpdateCheckRunRetryCount,
+	})
+
+	err := workflow.ExecuteActivity(ctx, p.Activities.UpdateCheckRun, activities.UpdateCheckRunRequest{
+		Title:   terraform.BuildCheckRunTitle(deployRequest.Root.Name),
+		State:   state,
+		Repo:    deployRequest.Repo,
+		ID:      deployRequest.CheckRunID,
+		Summary: summary,
+	}).Get(ctx, nil)
+	if err != nil {
+		logger.Error(ctx, "unable to update check run", err.Error())
+	}
+}
+
+// For merged deployments, notify user of a force apply lock status and lock future deployments until signal is received
+func (p *RevisionProcessor) lock(ctx workflow.Context, deploymentInfo terraform.DeploymentInfo) error {
+	// We won't lock a manually triggered root
+	if deploymentInfo.Root.Trigger == root.ManualTrigger {
```

Comment:
If i understand it correctly, we only lock the root if the FA PR has not been merged to the main branch before any other deploy requests (`hence diverged`)? Can we move this check to the main function before getting commit direction since we don't need that info to evaluate this? And, it'll also make FA PRs agnostic of the commit direction `(if we plan to support FA on revisions behind the main branch too)`?

Reply:
I might be misunderstanding what you mean here, but yes we lock the root if the revision is in a diverged state. We can't move this to before getting the commit direction since we need to know if it is diverged right?